### PR TITLE
fix assign node to replica in nodeUp

### LINKED ISSUE #22231: [Bug]: [ResourceGroup] Search fails even though there are replica available
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20230215-9346f175
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus-2.3.0.dev34 
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

There are two replicas of this collection in two resource groups, one of which is unavailable due to scale-in querynode and insufficient memory, and the other is available. But the search returns exception
```
[2023-02-16 09:06:53,206 - ERROR - fouram]: RPC error: [search], <MilvusException: (code=1, message=fail to search on all shard leaders, err=fail to Search, QueryNode ID=4, reason=Search 8 failed, reason QueryNode 42 can't serve, recovering: target node id not match target id = 8, node id = 42 err %!w(<nil>))>, <Time:{'RPC start': '2023-02-16 09:06:53.173397', 'RPC error': '2023-02-16 09:06:53.206773'}> (decorators.py:108)
```
1. deploy a cluster with 4 querynodes
2. create collection and insert 2m data, create index
3. create a resource group `RG_0` and transfer 2 querynodes into `RG_0` from default RG
4. load collection with 2 replicas into 2 RGs
```
collection.load(collName, replica_number=2, _resource_groups=['RG_0', '__default_resource_group']
```
5. search successfully
6. scale-in 1 querynodes (only 3 querynodes left)
7. The RG with offline qn does cannot balance the segments to the online qn due to insufficient memory, so one of the replica is unavailable. 
8. Search failed

### Expected Behavior

The search is expected to succeed because one of the replicas is available

### Steps To Reproduce

```markdown
argo workflow name: fouram-scale-9svxt
run the case test_scale_in_qns_memory in fouram repo. contact me for specific questions.
```

### Milvus Log

You can search for logs in Loki. if you want the raw log file for some components, tell me.
cluster: devops
namespace: chaos-testing
server pods:
```
fouram-scale-9svxt-op-63-4055-milvus-datacoord-5fb48b7cd7-qmbwj     Running     0            2m      10.102.7.213      devops-node11     
fouram-scale-9svxt-op-63-4055-milvus-datanode-9b689864-pktwt        Running     0            2m      10.102.10.15      devops-node20     
fouram-scale-9svxt-op-63-4055-milvus-indexcoord-97cd4888f-7tb5x     Running     0            2m      10.102.7.215      devops-node11     
fouram-scale-9svxt-op-63-4055-milvus-indexnode-5f5f694dd4-dv89k     Running     0            2m      10.102.7.214      devops-node11     
fouram-scale-9svxt-op-63-4055-milvus-indexnode-5f5f694dd4-vwlkc     Running     0            2m      10.102.9.13       devops-node13     
fouram-scale-9svxt-op-63-4055-milvus-proxy-758b57df58-9vbz2         Running     0            2m      10.102.9.12       devops-node13     
fouram-scale-9svxt-op-63-4055-milvus-querycoord-7c7b75765ccvnr9     Running     0            2m      10.102.5.153      devops-node21     
fouram-scale-9svxt-op-63-4055-milvus-querynode-77c44d5ff6-2qlwx     Running     0            2m      10.102.5.156      devops-node21     
fouram-scale-9svxt-op-63-4055-milvus-querynode-77c44d5ff6-888g8     Running     0            2m      10.102.7.216      devops-node11     
fouram-scale-9svxt-op-63-4055-milvus-querynode-77c44d5ff6-bzrkq     Running     0            2m      10.102.6.204      devops-node10     
fouram-scale-9svxt-op-63-4055-milvus-querynode-77c44d5ff6-jrklj     Running     0            2m      10.102.10.16      devops-node20     
fouram-scale-9svxt-op-63-4055-milvus-rootcoord-5c766bb9dc-krj7r     Running     0            2m      10.102.7.212      devops-node11     
fouram-scale-9svxt-op-63-4055-etcd-0                                Running     0            6m      10.102.5.138      devops-node21     
fouram-scale-9svxt-op-63-4055-etcd-1                                Running     0            6m      10.102.7.198      devops-node11     
fouram-scale-9svxt-op-63-4055-etcd-2                                Running     0            6m      10.102.10.3       devops-node20     
fouram-scale-9svxt-op-63-4055-pulsar-bookie-0                       Running     0            6m      10.102.7.203      devops-node11     
fouram-scale-9svxt-op-63-4055-pulsar-bookie-1                       Running     0            6m      10.102.6.200      devops-node10     
fouram-scale-9svxt-op-63-4055-pulsar-bookie-2                       Running     0            6m      10.102.5.147      devops-node21     
fouram-scale-9svxt-op-63-4055-pulsar-broker-0                       Running     0            6m      10.102.5.131      devops-node21     
fouram-scale-9svxt-op-63-4055-pulsar-proxy-0                        Running     0            6m      10.102.10.254     devops-node20     
fouram-scale-9svxt-op-63-4055-pulsar-recovery-0                     Running     0            6m      10.102.5.130      devops-node21     
fouram-scale-9svxt-op-63-4055-pulsar-zookeeper-0                    Running     0            6m      10.102.5.136      devops-node21     
fouram-scale-9svxt-op-63-4055-pulsar-zookeeper-1                    Running     0            6m      10.102.7.209      devops-node11     
fouram-scale-9svxt-op-63-4055-pulsar-zookeeper-2                    Running     0            5m      10.102.9.11       devops-node13     
fouram-scale-9svxt-op-63-4055-minio-0                               Running     0            6m      10.102.5.137      devops-node21     
fouram-scale-9svxt-op-63-4055-minio-1                               Running     0            6m      10.102.10.5       devops-node20     
fouram-scale-9svxt-op-63-4055-minio-2                               Running     0            6m      10.102.7.207      devops-node11     
fouram-scale-9svxt-op-63-4055-minio-3                               Running     0            6m      10.102.9.8        devops-node13     
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>
fix #22231

### CODE REVIEW

File: internal/querycoordv2/server.go [hunk 1]
```diff
@@ -648,28 +648,32 @@ func (s *Server) handleNodeUp(node int64) {
...
-			replica := replicas[0]
-			// TODO(yah01): this may fail, need a component to check whether a node is assigned
-			err = s.meta.ReplicaManager.AddNode(replica.GetID(), node)
-			if err != nil {
-				log.Warn("failed to assign node to replicas",
-					zap.Int64("replicaID", replica.GetID()),
-					zap.Error(err),
-				)
+	rgs := s.meta.ResourceManager.ListResourceGroups()
+	if len(rgs) == 1 {
```
Comment (yah01) [inline internal/querycoordv2/server.go:652]: Move these to resource manager may make sense

Comment (weiliu1031) [reply internal/querycoordv2/server.go:652]: > Move these to resource manager may make sense

resource manager doesn't care about how to assign node to replica

File: internal/querycoordv2/observers/resource_observer.go [hunk 2]
```diff
@@ -93,14 +94,43 @@ func (ob *ResourceObserver) checkResourceGroup() {
...
+						if replica == nil {
+							replicas := ob.meta.ReplicaManager.GetByCollectionAndRG(collection, rgName)
+							if len(replicas) == 0 {
+								continue
+							}
+							sort.Slice(replicas, func(i, j int) bool {
+								return replicas[i].Len() < replicas[j].Len()
+							})
+							replica := replicas[0]
+							// TODO(yah01): this may fail, need a component to check whether a node is assigned
```
[the remark this answers is not in the document]
Comment (weiliu1031) [reply internal/querycoordv2/observers/resource_observer.go:119]: not yet

File: internal/querycoordv2/meta/resource_manager.go [hunk 3]
```diff
@@ -466,15 +481,36 @@ func (rm *ResourceManager) HandleNodeDown(node int64) (string, error) {
...
-		return rgName, rm.groups[rgName].unassignNode(node, 0)
+		return "", err
 	}
 
-	return "", ErrNodeNotAssignToRG
+	log.Info("HandleNodeDown: remove node from resource group",
+		zap.String("rgName", rgName),
+		zap.Int64("node", node),
+	)
+	return rgName, rm.groups[rgName].unassignNode(node, 0)
```
[the remark this answers is not in the document]
Comment (weiliu1031) [reply internal/querycoordv2/meta/resource_manager.go:513]: no, we choose to modify etcd first, if save to etcd failed, nothing more need to do. and `unassignNode` won't failed, so it's safe here

**Review by jiaoew1991 — APPROVED** [review]
/lgtm

**Pull request merged**